### PR TITLE
Feat/issue 39 use a more reliable source of prices

### LINKED ISSUE #39: Use a more reliable source of prices
On Coignecko:
```
Data granularity is automatic (cannot be adjusted)
1 day from current time = 5 minute interval data
1 - 90 days from current time = hourly data
above 90 days from current time = daily data (00:00 UTC)
```

look into:
- CryptoCompare

### PR DESCRIPTION
**Pull request by IlluvatarEru**
## What does this PR do?

Adding Uniswap as a new source of prices

## Related issue

Closes #39 
## Testing

Ran on VPS.

## Checklist before merging
- [x] Read the [contributing guide](https://github.com/flashbots/mev-inspect-py/blob/main/CONTRIBUTING.md)
- [x] Installed and ran pre-commit hooks
- [x] All tests pass with `./mev test`

### CODE REVIEW

File: profit_analysis/prices.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        await pricer.create(token_address)
+        blocks = [
+            block_number_min + i
+            for i in range(int(block_number_max + 1 - block_number_min))
+        ]
+        print(blocks)
+        tasks = []
+        max_c = 1
+        max_concurrency = asyncio.Semaphore(max_c)
+        for block in range(block_number_min, block_number_max, max_c):
```
Comment (IlluvatarEru) [inline profit_analysis/prices.py]: there must be something wrong around here. Getting
`maximum recursion depth exceeded while calling a Python object` error

File: profit_analysis/prices.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+async def safe_get_price(pricer, block, max_concurrency_semaphore):
+    async with max_concurrency_semaphore:
+        return await pricer.get_price_at_block(block)
+
+
+async def get_uniswap_historical_prices(
+    target_blocks,
+    token_address,
+    chain=POLYGON_CHAIN,
+    max_concurrency=1,
```
Comment (IlluvatarEru) [inline profit_analysis/prices.py]: What are optimal values here for `max_concurrency` and `block_batch_size`? @0xpanoramix

File: profit_analysis/prices.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    target_blocks = [int(b) for b in target_blocks]
+    pricer = UniswapPricer(W3, chain)
+    # we use USDC as a base token
+    await pricer.create(token_address)
+    tasks = []
+    max_concurrency_semaphore = asyncio.Semaphore(max_concurrency)
+
+    for start_block_number_index in range(0, len(target_blocks), block_batch_size):
+        start_block_number = target_blocks[start_block_number_index]
+        print(f"start_block_number={start_block_number}")
```
Comment (IlluvatarEru) [inline profit_analysis/prices.py]: - [x] Remove `print` once we have chosen the right values for above parameters and tested it did not error

File: mev_inspect/block.py [hunk 4]
```diff
@@ -25,6 +25,7 @@
 
 async def _get_logs_for_topics(after_block, before_block, topics):
     trials = 0
+    n_trials = 3
     while trials < 3:
```
Comment (0xpanoramix) [inline mev_inspect/block.py]: Consider using your newly created variable here:

```suggestion
    while trials < n_trials:
```

File: mev_inspect/block.py [hunk 5]
```diff
@@ -137,18 +138,19 @@ async def classify_logs(logs, pool_reserves):
 
 async def get_pool_reserves(addr, pool_address):
     trials = 0
+    n_trials = 3
     while trials < 3:
```
Comment (0xpanoramix) [inline mev_inspect/block.py]: Same thing here:

```suggestion
    while trials < n_trials:
```

File: profit_analysis/block_utils.py [hunk 6]
```diff
@@ -19,6 +19,7 @@ def add_block_timestamp(profit_by_block):
 
 def get_block_timestamp(block):
     trials = 0
+    n_trials = 3
     while trials < 3:
```
Comment (0xpanoramix) [inline profit_analysis/block_utils.py]:
```suggestion
    while trials < n_trials:
```

**Review by 0xpanoramix — APPROVED** [review]

File: mev_inspect/block.py (hunk 4 above)
Comment (IlluvatarEru) [reply mev_inspect/block.py]: :heavy_check_mark:

File: profit_analysis/block_utils.py (hunk 6 above)
Comment (IlluvatarEru) [reply profit_analysis/block_utils.py]: :heavy_check_mark:

File: mev_inspect/block.py (hunk 5 above)
Comment (IlluvatarEru) [reply mev_inspect/block.py]: :heavy_check_mark:

File: profit_analysis/prices.py (hunk 2 above)
Comment (0xpanoramix) [reply profit_analysis/prices.py]: Max concurrency -> 10 is enough, we'll update later on is required.
Block batch size -> 1024, we'll use the stride to properly manage the number of logs retrieved at each request.

Comment (IlluvatarEru) [reply profit_analysis/prices.py]: Thanks :white_check_mark:

**Pull request merged**